runtime/near-evm-runner/src/utils.rs

```diff
@@ -183,66 +187,497 @@ pub fn format_log(topics: Vec<H256>, data: &[u8]) -> std::result::Result<Vec<u8>
+pub struct Method {
+    pub name: String,
+    pub raw: String,
+    pub args: Vec<Arg>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct MethodAndTypes {
+    pub method: Method,
+    pub type_sequences: Vec<String>,
+    pub types: HashMap<String, Method>,
+}
+
+impl Arg {
+    fn parse(text: &str) -> Result<(Arg, &str)> {
+        let (type_raw, remains) = parse_type_raw(text)?;
+        let t = parse_type(&type_raw)?;
+        let remains = consume(remains, ' ')?;
+        let (name, remains) = parse_ident(remains)?;
+        Ok((Arg { type_raw, t, name }, remains))
+    }
+
+    fn parse_args(text: &str) -> Result<(Vec<Arg>, &str)> {
+        let mut remains = consume(text, '(')?;
+        if remains.len() == 0 {
+            return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+        }
+        let mut args = vec![];
+        let first = remains.chars().next().unwrap();
+        if is_arg_start(first) {
+            let (arg, r) = Arg::parse(remains)?;
+            remains = r;
+            args.push(arg);
+            while remains.chars().next() == Some(',') {
+                remains = consume(remains, ',')?;
+                let (arg, r) = Arg::parse(remains)?;
+                remains = r;
+                args.push(arg);
+            }
+        }
+
+        let remains = consume(remains, ')')?;
+
+        Ok((args, remains))
+    }
+}
+
+impl Method {
+    fn parse(method_def: &str) -> Result<(Method, &str)> {
+        let (name, remains) = parse_ident(method_def)?;
+        let (args, remains) = Arg::parse_args(remains)?;
+        Ok((
+            Method { name, args, raw: method_def[..method_def.len() - remains.len()].to_string() },
+            remains,
+        ))
+    }
+}
+
+impl MethodAndTypes {
+    pub fn parse(method_def: &str) -> Result<Self> {
+        let mut method_def = method_def;
+        let mut parsed_types = HashMap::new();
+        let mut type_sequences = vec![];
+        let (method, mut types) = Method::parse(method_def)?;
+        while types.len() > 0 {
+            let (ty, remains) = Method::parse(types)?;
+            type_sequences.push(ty.name.clone());
+            parsed_types.insert(ty.name.clone(), ty);
+            types = remains;
+        }
+        Ok(MethodAndTypes { method, types: parsed_types, type_sequences })
+    }
+}
+
+fn parse_ident(text: &str) -> Result<(String, &str)> {
+    let mut chars = text.chars();
+    if text.len() == 0 || !is_arg_start(chars.next().unwrap()) {
+        return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+    }
+
+    let mut i = 1;
+    for c in chars {
+        if !is_arg_char(c) {
+            break;
+        }
+        i += 1;
+    }
+    Ok((text[..i].to_string(), &text[i..]))
+}
+
+fn parse_type_raw(text: &str) -> Result<(String, &str)> {
+    let i =
+        text.find(' ').ok_or(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName))?;
+    Ok((text[..i].to_string(), &text[i..]))
```

Comment:
Looks like a `split` method

Reply:
split doesn't work because want only split on first space

Reply:
I think `splitn` is what you are looking for, it allows limiting how many sub-strings are returned.

https://doc.rust-lang.org/std/primitive.str.html#method.splitn

Reply:
I just learned splitn, thanks! but it still has a problem of allowing "up
to n" not "exactly n" so it isn't easier than find and split (need add
```
if splitted.len() != 2 {
  return Err(...)
}
```

Reply:
Right, that makes sense. You could also do something like

```
let mut iter = text.splitn(2, ' ');
let (prefix, suffix) = iter.next().and_then(|p| iter.next().map(|s| (p, s))).ok_or(...);
Ok((prefix.to_string(), suffix))
```

But I don't think that is cleaner than your solution with `find`; the mapping over options might even be less clear.

As a side note there is `split_once`, but it is not yet stabilized and probably a small thing like this is not worth adding a nightly feature over.
https://doc.rust-lang.org/std/primitive.str.html#method.split_once